anonymizer/OpenDiffix.Service/AggregationHooks.fs

```diff
@@ -0,0 +1,127 @@
+module AggregationHooks
+
+open OpenDiffix.Core
+
+module BucketAttributes =
+  let IS_LED_MERGED = "is_led_merged"
+
+let private findSingleIndex cond arr =
+  arr
+  |> Array.indexed
+  |> Array.filter (snd >> cond)
+  |> function
+    | [| index, _item |] -> Some index
+    | _ -> None
+
+let private findAggregator aggFn aggregators =
+  aggregators |> findSingleIndex (fun ((fn, _), _) -> fn = aggFn)
+
+let private isLowCount lowCountIndex bucket =
+  bucket.Aggregators.[lowCountIndex].Final(bucket.ExecutionContext)
+  |> Value.unwrapBoolean
+
+let private isGlobalAggregation (aggregationContext: AggregationContext) =
+  Array.isEmpty aggregationContext.GroupingLabels
+
+module Led =
+  type private SiblingPerColumn =
+    | NoBuckets
+    | SingleBucket of bucket: Bucket * isLowCount: bool
+    | MultipleBuckets
+
+  /// Compares 2 rows and returns the index where a single column is not equal.
+  /// Returns `None` if rows are equal or are different in more than one column.
+  let private findSingleNonMatchingColumn max (row1: Row) (row2: Row) =
+    let rec compare notEqualAt i =
+      if i = max then
+        // End of comparison, return matched index (may be None)
+        notEqualAt
+      else if row1.[i] = row2.[i] then
+        // Columns equal, try next one
+        compare notEqualAt (i + 1)
+      else if Option.isSome notEqualAt then
+        // A non-equal column already exists, abort
+        None
+      else
+        // Mark current index as not equal and keep going to verify no others are
+        compare (Some i) (i + 1)
+
+    compare None 0
+
+  let private anonymizingAggregatorIndexes (aggregationContext: AggregationContext) =
+    aggregationContext.Aggregators
+    |> Seq.map fst
+    |> Seq.indexed
+    |> Seq.choose (fun (i, aggSpec) -> if Aggregator.isAnonymizing aggSpec then Some i else None)
+    |> Seq.toArray
+
+  // For LED hook we merge only anonymizing aggregators.
+  let private mergeGivenAggregatorsInto (targetBucket: Bucket) aggIndexes (sourceBucket: Bucket) =
+    let sourceAggregators = sourceBucket.Aggregators
+    let targetAggregators = targetBucket.Aggregators
+
+    aggIndexes
+    |> Array.iter (fun i -> targetAggregators.[i].Merge(sourceAggregators.[i]))
+
+  let private led lowCountIndex (aggregationContext: AggregationContext) (buckets: Bucket seq) =
+    let groupingLabelsLength = aggregationContext.GroupingLabels.Length
+    let anonAggregators = anonymizingAggregatorIndexes aggregationContext
+
+    let buckets =
+      buckets
+      |> Seq.map (fun bucket -> bucket, isLowCount lowCountIndex bucket)
+      |> Seq.toArray
+
+    buckets
+    |> Array.iteri (fun i (victimBucket, isLowCount) ->
+      if isLowCount then
```

Comment:
I plan on rewriting LED so this will be obsolete soon. For now just know that this checks for some very specific conditions which won't happen 99% of the time. If a victim bucket is found under those conditions, it is merged somewhere else to "hide in the crowd".

You should only worry about `module StarBucket` right now because the `Led` code will change and it's not worth knowing the implementation details right now.